File: app/src/main/java/com/example/android/securenote/GetPasswordDialog.java

package com.example.android.securenote;

import android.app.Activity;
import android.os.Bundle;
import android.text.Editable;
import android.text.TextWatcher;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.View.OnClickListener;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.EditText;

import androidx.fragment.app.DialogFragment;

public class GetPasswordDialog extends DialogFragment implements
        OnClickListener, TextWatcher {
    private static final String TAG = GetPasswordDialog.class.getSimpleName();

    private static final String VERIFY_PASSWORD_REQUEST_PARAM = "verifyPassword";
    private static final String MIN_PASSWORD_LENGTH_REQUEST_PARAM = "minPasswordLength";
    private static final String REQUEST_PARAM = "requestType";

    static GetPasswordDialog newInstance(int requestType,
                                         int minPasswordLength,
                                         boolean verifyPassword) {
        GetPasswordDialog dialog = new GetPasswordDialog();
        Bundle args = new Bundle();
        args.putBoolean(VERIFY_PASSWORD_REQUEST_PARAM, verifyPassword);
        args.putInt(MIN_PASSWORD_LENGTH_REQUEST_PARAM, minPasswordLength);
        args.putInt(REQUEST_PARAM, requestType);
        dialog.setArguments(args);

        return dialog;
    }

    public interface OnPasswordListener {
        void onPasswordValid(int requestType, String password);

        void onPasswordCancel();
    }

    private EditText password;
    private EditText passwordVerification;
    private Button okButton;
    private int minPasswordLength;

    private OnPasswordListener passwordListener;

    @Override
    public void onAttach(Activity activity) {
        super.onAttach(activity);
        try {
            passwordListener = (OnPasswordListener) activity;
        } catch (ClassCastException e) {
            throw new IllegalArgumentException(
                    activity.getClass().getSimpleName()
                            + " should implement OnPasswordListener");
        }
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        getDialog().setTitle(R.string.get_password_label);
        View content = inflater.inflate(R.layout.get_password, container, false);

        password = content.findViewById(R.id.password_text);
        passwordVerification = content.findViewById(R.id.password_verification_text);
        okButton = content.findViewById(R.id.ok_button);

        content.findViewById(R.id.cancel_button).setOnClickListener(this);
        okButton.setOnClickListener(this);


        return content;
    }

    @Override
    public void onActivityCreated(Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        final Bundle args = getArguments();

        boolean verifyPassword = args.getBoolean(VERIFY_PASSWORD_REQUEST_PARAM, true);
        if (verifyPassword) {
            okButton.setEnabled(false);
            minPasswordLength = args.getInt(MIN_PASSWORD_LENGTH_REQUEST_PARAM, 0);
            if (minPasswordLength > 0) {
                password.setHint(super.getString(R.string.password_hint_min_length,
                        minPasswordLength));
            }
            password.addTextChangedListener(this);
            passwordVerification.addTextChangedListener(this);
        } else {
            passwordVerification.setVisibility(View.GONE);
        }
    }

    @Override
    public void onPause() {
        super.onPause();
        password.getText().clear();
        passwordVerification.getText().clear();
        Log.d(TAG, "Cleared password fields");
    }

    public void onClick(View v) {
        switch (v.getId()) {
            case R.id.ok_button:
                final int requestType = getArguments().getInt(REQUEST_PARAM);
                final String password = this.password.getText().toString();
                passwordListener.onPasswordValid(requestType, password);
                break;
            case R.id.cancel_button:
                passwordListener.onPasswordCancel();
                break;
            default:
                throw new IllegalArgumentException("Invalid Button");
        }
        // the passwords will be cleared during onPause()
        dismiss();
    }

    public void afterTextChanged(Editable s) {
        if (password.length() < minPasswordLength) {
            Log.d(TAG, "Password too short");
            okButton.setEnabled(false);
        } else if (password.length() != passwordVerification.length()) {
            Log.d(TAG, "Passwords' length differs");
            okButton.setEnabled(false);
        } else {
            for (int i = 0; i < password.getText().length(); i++) {
                if (password.getText().charAt(i) != passwordVerification.getText()
                        .charAt(i)) {
                    Log.d(TAG, "Passwords differ");
                    okButton.setEnabled(false);
                    return;
                }
            }
            Log.d(TAG, "Passwords are the same");
            okButton.setEnabled(true);
        }
    }

    public void beforeTextChanged(CharSequence s, int start, int count, int after) {
        // ignored
    }

    public void onTextChanged(CharSequence s, int start, int before, int count) {
        // ignored
    }
}

File: app/src/main/java/com/example/android/securenote/GetPasswordDialog.kt
package com.example.android.securenote

import android.content.Context
import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.View.OnClickListener
import android.view.ViewGroup
import com.example.android.securenote.databinding.GetPasswordBinding


class GetPasswordDialog : androidx.fragment.app.DialogFragment(), OnClickListener, TextWatcher {
    private var minPasswordLength: Int = 0
    private var passwordListener: OnPasswordListener? = null
    private lateinit var getPasswordBinding: GetPasswordBinding

    interface OnPasswordListener {
        fun onPasswordValid(requestType: Int, password: String)

        fun onPasswordCancel()
    }

    override fun onAttach(context: Context) {
        super.onAttach(context)
        try {
            passwordListener = context as OnPasswordListener
        } catch (e: ClassCastException) {
            throw IllegalArgumentException("Must implement OnPasswordListener")
        }
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        dialog?.setTitle(R.string.get_password_label)
        getPasswordBinding = GetPasswordBinding.inflate(layoutInflater)

        getPasswordBinding.cancelButton.setOnClickListener(this)
        getPasswordBinding.okButton.setOnClickListener(this)

        return getPasswordBinding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        val args = arguments ?: return

        val verifyPassword = args.getBoolean(VERIFY_PASSWORD_REQUEST_PARAM, true)
        if (verifyPassword) {
            getPasswordBinding.okButton.isEnabled = false
            minPasswordLength = args.getInt(MIN_PASSWORD_LENGTH_REQUEST_PARAM, 0)
            if (minPasswordLength > 0) {
                getPasswordBinding.passwordText.hint = super.getString(
                    R.string.password_hint_min_length,
                    minPasswordLength
                )
            }
            getPasswordBinding.passwordText.addTextChangedListener(this)
            getPasswordBinding.passwordVerificationText.addTextChangedListener(this)
        } else {
            getPasswordBinding.passwordVerificationText.visibility = View.GONE
        }
    }

    override fun onPause() {
        super.onPause()
        getPasswordBinding.passwordText.text.clear()
        getPasswordBinding.passwordVerificationText.text.clear()
        Log.d(TAG, "Cleared password fields")
    }

    override fun onClick(v: View) {
        when (v.id) {
            R.id.ok_button -> {
                val requestType = requireArguments().getInt(REQUEST_PARAM)
                val password = getPasswordBinding.passwordText.text.toString()
                passwordListener!!.onPasswordValid(requestType, password)
            }
            R.id.cancel_button -> passwordListener!!.onPasswordCancel()
            else -> throw IllegalArgumentException("Invalid Button")
        }
        // the passwords will be cleared during onPause()
        dismiss()
    }

    override fun afterTextChanged(s: Editable) {
        when {
            getPasswordBinding.passwordText.length() < minPasswordLength -> {
                Log.d(TAG, "Password too short")
                getPasswordBinding.okButton.isEnabled = false
            }
            getPasswordBinding.passwordText.length() !=
                    getPasswordBinding.passwordVerificationText.length() -> {
                Log.d(TAG, "Passwords' length differs")
                getPasswordBinding.okButton.isEnabled = false
            }
            else -> {
                for (i in getPasswordBinding.passwordText.text.indices) {
                    if (getPasswordBinding.passwordText.text[i] !=
                        getPasswordBinding.passwordVerificationText.text[i]
                    ) {
                        Log.d(TAG, "Passwords differ")
                        getPasswordBinding.okButton.isEnabled = false
                        return
                    }
                }
                Log.d(TAG, "Passwords are the same")
                getPasswordBinding.okButton.isEnabled = true
            }
        }
    }

    override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {
        // ignored
    }

    override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {
        // ignored
    }

    companion object {
        private val TAG = GetPasswordDialog::class.java.simpleName

        const val VERIFY_PASSWORD_REQUEST_PARAM = "verifyPassword"
        const val MIN_PASSWORD_LENGTH_REQUEST_PARAM = "minPasswordLength"
        const val REQUEST_PARAM = "requestType"

        fun newInstance(
            requestType: Int,
            minPasswordLength: Int,
            verifyPassword: Boolean
        ): GetPasswordDialog {
            val dialog = GetPasswordDialog()
            val args = Bundle()
            args.putBoolean(VERIFY_PASSWORD_REQUEST_PARAM, verifyPassword)
            args.putInt(MIN_PASSWORD_LENGTH_REQUEST_PARAM, minPasswordLength)
            args.putInt(REQUEST_PARAM, requestType)
            dialog.arguments = args

            return dialog
        }
    }
}
